bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/utils/flowcontrol/FlowController.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.mirroring.hbase1_x.utils.flowcontrol;
+
+import com.google.api.core.InternalApi;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+/**
+ * FlowController limits the number of concurrently performed requests to the secondary database.
+ * Call to {@link #asyncRequestResource(RequestResourcesDescription)} returns a future that will be
+ * completed when {@link FlowControlStrategy} decides that it can be allowed to perform the
+ * requests. The future might also be completed exceptionally if the resource was not allowed to
+ * obtain the resources.
+ *
+ * <p>Order of allowing requests in determined by {@link FlowControlStrategy}.
+ *
+ * <p>Thread-safe.
+ */
+@InternalApi("For internal usage only")
+public class FlowController {
```

Comment:
What is the split of responsiblities between FlowController and FlowControlStrategy? I am not sure what we achieve  by have these 2 different entities? You have have a flow control interface and implement various flow controllers?

Reply:
This is also an option, but currently FlowController handles matching Reservation objects with Resources and handling cancellation.